File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.js
function solution(edges) {
  // 해결 방안
  // input 으로 그래프 형성
  // 그래프가 각 어떤 것인지 판단
  // [정점 번호, 도넛, 막대, 8자] 답 양식
  var answer = new Array(4).fill(0);

  // 그래프 생성 : { node: [진출차수, 진입차수]}
  const graph = {};
  for (let i = 0; i < edges.length; i++) {
    let [nodeStart, nodeEnd] = edges[i];
    // 노드를 기준으로 진출, 진입 차수 판단
    // 따라서 시작, 끝 점 모두 key 값으로 있는지 판단해야함.
    // graph 에 해당 노드 정보가 없으면 초기화
    if (!(nodeStart in graph)) {
      graph[nodeStart] = [0, 0];
    }
    if (!(nodeEnd in graph)) {
      graph[nodeEnd] = [0, 0];
    }
    graph[nodeStart][0]++;
    graph[nodeEnd][1]++;
  }

  // created node
  for (const node in graph) {
    if (!graph[node][1] && graph[node][0] >= 2) {
      answer[0] = parseInt(node);
    }
  }

  // total graph num
  const total = graph[answer[0]][0];

  // 생성 정점에서 그래프와의 연결 제거
  edges.forEach((edge) => {
    const [s, e] = edge;
    if (s !== answer[0]) return;
    graph[e][1]--;
  });

  for (const node in graph) {
    const [o, i] = graph[node];
    if (o === 2 && i === 2) answer[3]++;
    if (i === 0 && o <= 1) answer[2]++;
  }

  answer[1] = total - (answer[3] + answer[2]);

  return answer;
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
def solution(edges):
    answer = [0] * 4 # [생성 정점 번호, 도넛, 막대, 8자]
    
    graph = {} # {node: [진출차수, 진입차수]}
    for edge in edges:
        s, e = edge
        if s not in graph:
            graph[s] = [0, 0]
        if e not in graph:
            graph[e] = [0, 0]
        graph[s][0] += 1
        graph[e][1] += 1
    
    # 생성 정점은 진출차수 2이상, 진입차수 0
    for key in graph:
        if graph[key][0] >= 2 and not graph[key][1]:
            answer[0] = int(key)
            
    # 총 그래프 수: 생성 정점에서 진출 차수
    graph_num = graph[answer[0]][0]
    
    # 생성 정점을 아니까 그래프 연결 간선 제거
    for edge in edges:
        s, e = edge
        if s == answer[0]: graph[e][1] -= 1
        else: continue
        
    # 각 그래프는 이제 고유의 형태를 유지하고 있음
    # 8자는 중심인 정점이 진입 2, 진출 2
    # 막대는 중심인 정점이 진입 0, 진출 <= 1
    for key in graph:
        o, i = graph[key]
        if o == 2 and i == 2: answer[3] += 1
        if o <= 1 and i == 0: answer[2] += 1        
        
    # 총 그래프 수에서 8자, 막대 개수 빼주면 도넛 그래프의 개수
    answer[1] = graph_num - (answer[2] + answer[3])
    
    return answer